include/robot_interfaces/sensors/pylon_driver.hpp

```diff
@@ -31,27 +31,66 @@ class PylonDriver : public SensorDriver<CameraObservation>
     Pylon::PylonAutoInitTerm auto_init_term_;
     Pylon::CInstantCamera camera_;
     Pylon::CTlFactory& tl_factory_;
-    Pylon::DeviceInfoList_t device_list_;
-    Pylon::DeviceInfoList_t::const_iterator device_iterator_;
     Pylon::CImageFormatConverter format_converter_;
     Pylon::CPylonImage pylon_image_;
-    int num_devices_;
+    const std::string& device_user_id_to_open = "cam_1";
 
     PylonDriver()
         : tl_factory_(Pylon::CTlFactory::GetInstance())
     {
-        device_iterator_ = device_list_.begin();
-        camera_.Attach(tl_factory_.CreateDevice(*device_iterator_));
-        camera_.Open();
-        camera_.MaxNumBuffer = 5;
-        format_converter_.OutputPixelFormat = Pylon::PixelType_BGR8packed;
+        Pylon::PylonInitialize();
+        Pylon::DeviceInfoList_t device_list;
 
-        camera_.StartGrabbing();
+        if (tl_factory_.EnumerateDevices(device_list) == 0)
+        {
+            std::cerr << "No devices present, please connect one." << std::endl;
+            Pylon::PylonTerminate();
+        }
+        else
+        {
+            Pylon::DeviceInfoList_t::const_iterator device_iterator;
+            bool found_desired_device = false;
+
+            for(device_iterator = device_list.begin(); device_iterator != device_list.end(); ++device_iterator)
+            {
+                std::string device_user_id_found(device_iterator->GetUserDefinedName());
+                if((0 == device_user_id_to_open.compare(device_user_id_found)) ||
+                (device_user_id_to_open.length() < device_user_id_found.length() &&
+                (0 == device_user_id_found.compare(device_user_id_found.length() -
+                                                    device_user_id_to_open.length(),
+                                                    device_user_id_to_open.length(),
+                                                    device_user_id_to_open) )
+                )
+                )
+                {
+                    found_desired_device = true;
+                    break;
+                }
+            }
+
+            if (found_desired_device)
+            {
+                camera_.Attach(tl_factory_.CreateDevice(*device_iterator));
+            }
+            else
+            {
+                device_iterator = device_list.begin();
+                camera_.Attach(tl_factory_.CreateDevice(*device_iterator));
+                std::cout << "Desired device not found. Creating a camera object with the first device id in device list." << std::endl;                
```

Comment:
I would rather throw an exception if the desired camera is not found.
However, I think it would be nice to add a case like this (i.e. using the first camera) in case the user sets an empty string as id.